### PR TITLE
Allow downloading computed isosurfaces as stl plus re-upload

### LINKED ISSUE #3714: Download isosurfaces as stl

### PR DESCRIPTION
**Pull request by philippotto**
This PR adds a download button to the mesh tab which will start an stl download for the currently active cell id (shift+click to change; only works in view mode).
Within the stl file, the header is used to decode that the file represents an computed isosurface plus the cell id. This information is used to deal with the STL import differently. The differences are:

- for a normal mesh import, we persist the STL on the server, show the element in a list which allows further modification (position, visibility, name etc.). also, a normal mesh doesn't have a designated color
- for an isosurface mesh, we simply add the mesh to the scene as if it were computed from the server. the only difference is where the mesh is coming from and that we don't update the cache which avoids redundant re-computations. the cache is indexed by the position for which isosurfaces were requested. however, I didn't want to encode these within the stl (the header size is limited). I think, that's fine for now. the only downside is that it **can** happen that the isosurface is redundantly recomputed (however, they will perfectly overlap so it's not a rendering issue).

All in all, I'm not really satisfied with the current state of meshes and isosurfaces. UI- and code-wise it feels quite messy. A proper clean up would be good. But probably out of scope for this sprint (requires a bit of brainstorming, too, I think).

### URL of deployed dev instance (used for testing):
- https://stldownload.webknossos.xyz

### Steps to test:
- open a dataset in view mode with a segmentation
- enable isosurfaces rendering
- shift click on a cell (right now, the center position is used as result of another bug)
- click the download button in the mesh view
- refresh the page
- use the import button in the mesh view to import the stl you just downloaded
- the mesh should be rendered as it was before the page refresh (same position, color & appearance)

### Issues:
- fixes #3714 

------
- [ ] Updated [changelog](../blob/master/CHANGELOG.md#unreleased)
- [X] Ready for review

### CODE REVIEW

File: frontend/javascripts/libs/error_handling.js [hunk 1]
```diff
@@ -132,7 +132,7 @@ class ErrorHandling {
   notify(error: Error, optParams: Object = {}, escalateToSlack: boolean = false) {
     const actionLog = getActionLog();
     this.airbrake.notify({ error, params: { ...optParams, actionLog } });
-    if (escalateToSlack) {
+    if (false && escalateToSlack) {
```
Comment (daniel-wer) [inline frontend/javascripts/libs/error_handling.js]: We could use this PR to remove this altogether or add a TODO :)

File: frontend/javascripts/libs/stl_exporter.js [hunk 2]
```diff
@@ -0,0 +1,159 @@
+// @noflow
```
Comment (daniel-wer) [inline frontend/javascripts/libs/stl_exporter.js]: This code was copied (nearly) as is, right? Maybe add a link to the original source, this way we could also check for newer versions if there are bugs or if we need additional information :)

File: frontend/javascripts/oxalis/controller/scene_controller.js [hunk 3]
```diff
@@ -60,8 +60,12 @@ class SceneController {
...
   scene: THREE.Scene;
   rootGroup: THREE.Object3D;
-  stlMeshes: { [key: string]: THREE.Mesh };
-  isosurfacesGroup: THREE.Group;
+  stlMeshes: { [key: string]: THREE.Mesh } = {};
+
+  // isosurfacesRootGroup holds lights and one group per segmentation id.
+  // Each group can hold multiple meshes.
+  isosurfacesRootGroup: THREE.Group;
+  isosurfacesGroupsPerSegmentationId: { [key: number]: THREE.Group } = {};
```
Comment (daniel-wer) [inline frontend/javascripts/oxalis/controller/scene_controller.js]: What's the benefit of having a group per segmentation id?

File: frontend/javascripts/oxalis/model/sagas/handle_mesh_changes.js [hunk 4]
```diff
@@ -91,6 +91,7 @@ function* createMeshFromBuffer(action: CreateMeshFromBufferAction): Saga<void> {
   // Parse and persist STL in parallel
   const [geometry, meshMetaData] = yield _all([
     _call(parseStlBuffer, action.buffer),
+    _call(() => ({ position: [0, 0, 0], id: 3 })),
```
Comment (daniel-wer) [inline frontend/javascripts/oxalis/model/sagas/handle_mesh_changes.js]: Was this for testing?

File: frontend/javascripts/oxalis/model/sagas/isosurface_saga.js [hunk 5]
```diff
@@ -171,8 +197,35 @@ function* maybeLoadIsosurface(
...
+  const sceneController = getSceneController();
+  const geometry = sceneController.getIsosurfaceGeometry(currentIsosurfaceCellId);
+
+  const stl = exportToStl(geometry);
+
+  // Encode isosurface and cell id property
+  binaryIsosurfaceMarker.forEach((marker, index) => {
+    stl.setUint8(index, marker);
+  });
+  stl.setUint32(3, currentIsosurfaceCellId, true);
```
Comment (daniel-wer) [inline frontend/javascripts/oxalis/model/sagas/isosurface_saga.js]: The `3` should be extracted from this and the next method, maybe in a similar way as the binaryIsosurfaceMarker :)

File: frontend/javascripts/oxalis/controller/scene_controller.js (hunk 3 above)
Comment (daniel-wer) [reply frontend/javascripts/oxalis/controller/scene_controller.js]: Ah, I see, the download is per segment id :)

File: frontend/javascripts/oxalis/model/sagas/isosurface_saga.js [hunk 6]
```diff
@@ -1,22 +1,44 @@
 // @flow
+import { saveAs } from "file-saver";
```
Comment (daniel-wer) [inline frontend/javascripts/oxalis/model/sagas/isosurface_saga.js]: :+1:

File: frontend/javascripts/oxalis/view/right-menu/meshes_view.js [hunk 7]
```diff
@@ -130,6 +143,9 @@ class MeshesView extends React.Component<Props, { currentlyEditedMesh: ?MeshMeta
               Import
             </ButtonComponent>
           </Upload>
+          <ButtonComponent icon="download" onClick={this.props.downloadIsosurface}>
```
Comment (daniel-wer) [inline frontend/javascripts/oxalis/view/right-menu/meshes_view.js]: Should this include a note that only the isosurface of the currently active cell will be downloaded?

File: frontend/javascripts/oxalis/model/sagas/handle_mesh_changes.js (hunk 4 above)
Comment (philippotto) [reply frontend/javascripts/oxalis/model/sagas/handle_mesh_changes.js]: Yes, thanks for spotting this!

File: frontend/javascripts/libs/stl_exporter.js (hunk 2 above)
Comment (philippotto) [reply frontend/javascripts/libs/stl_exporter.js]: Good point, I added a link!

**Comment by philippotto** [comment]
As discussed in person, I piggybacked this PR to remove the slack-escalation completely.

**Pull request merged**